Fix Incorrect Scale Subresources and HPA e2e ScaleTargetRefs

DirectXMan12 opened the pull request:
The HPA e2es failed to actually set `apiVersion` on the created HPAs, which previous was ignored.  Since the polymorphic scale client was merged, this behavior is no longer tolerated (it was never correct to begin with, but it accidentally worked).

Additionally, the `apps` resources have their own version of scale.  Until `apps/v1beta1` and `apps/v1beta2` go away, we need to support those versions in the scale client.

Together, these broke some of the HPA e2es.

Fixes #54574

```release-note
NONE
```

DirectXMan12 committed:
[e2e] make sure to specify APIVersion in HPA tests

Previously, the HPA controller ignored APIVersion when resolving the
scale subresource for a kind, meaning if it was set incorrectly in the
HPA's scaleTargetRef, it would not matter.  This was the case for
several of the HPA e2e tests.

Since the polymorphic scale client merged, APIVersion now matters.  This
updates the HPA e2es to care about APIVersion, by passing kind as a full
GroupVersionKind, and not just a string.

DirectXMan12 commented:
cc @kubernetes/sig-autoscaling-pr-reviews @kubernetes/sig-autoscaling-test-failures @spiffxp 

cc kubernetes/sig-release#22

spiffxp commented:
/retest

DirectXMan12 commented:
cc @kubernetes/api-reviewers I assert that this isn't an API change, BTW, since the actual API object was an `extensions/v1beta1.Scale` (and, IIRC, discovery was correct) -- it was just the `apiVersion` field that was incorrect on the returned object.

liggitt commented on pkg/registry/apps/rest/storage_apps.go:
```diff
@@ -40,6 +41,15 @@ func (p RESTStorageProvider) NewRESTStorage(apiResourceConfigSource serverstorag
+		}
+		apiGroupInfo.SubresourceGroupVersionKind["deployments/scale"] = extensionsGroupVersion.WithKind("Scale")
+		apiGroupInfo.SubresourceGroupVersionKind["statefulsets/scale"] = extensionsGroupVersion.WithKind("Scale")
+		apiGroupInfo.SubresourceGroupVersionKind["replicasets/scale"] = extensionsGroupVersion.WithKind("Scale")
```
This won't work, for a few reasons:

1. this forces `Scale` to be uniform in all versions of the `apps` API group. apps/v1beta1 and apps/v1beta2 accidentally shipped their own versions of Scale. We're trying to fix that in apps/v1, but that requires per-version control over subresource versioning. It's likely we should remove `SubresourceGroupVersionKind` and start using the `GroupVersionKindProvider` interface that lets each storage indicate the external version of the subresource.

2. we shipped releases with the scale subresources in `apps/v1beta1` and `apps/v1beta2` returning and requiring `apps/v1beta1` and `apps/v1beta2` versions of Scale. If you submit an `extensions/v1beta1` Scale object to a 1.8 apps/v1beta2 scale subresource, you get:

```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "the API version in the data (extensions/v1beta1) does not match the expected API version (apps/v1beta2)",
  "reason": "BadRequest",
  "code": 400
}
```

We cannot break wire compatibility with shipped releases. We'll need to figure out how to keep apps/v1beta1 and apps/v1beta2 behavior as-is, and fix it up for apps/v1

liggitt commented:
> cc @kubernetes/api-reviewers I assert that this isn't an API change, BTW, since the actual API object was an extensions/v1beta1.Scale (and, IIRC, discovery was correct) -- it was just the apiVersion field that was incorrect on the returned object.

Unfortunately, we're committed to the wire format we returned and accepted for existing subresources in 1.8. That's the API currently exposed and written to.

> Additionally, the apps resources failed to properly set their subresource GVKs in storage, leading to Scales formatted as extensions/v1beta1.Scale being returned from the API server with their apiVersion set to apps/v2beta1, which yielded a scale version unknown to all.

That's not actually true. The storage handlers worked with extensions.Scale objects internally, but got converted to and from apps/v1beta1 and apps/v1beta2 Scale objects via conversion (autoConvert_extensions_Scale_To_v1beta1_Scale and autoConvert_extensions_Scale_To_v1beta2_Scale, etc)

liggitt commented:
Opened https://github.com/kubernetes/kubernetes/pull/54683 as an alternative to the storage changes, which gives us per-version control of the subresource GVK and uses the information returned by the storage consistently in discovery and endpoint registration

DirectXMan12 commented:
> That's not actually true. The storage handlers worked with extensions.Scale objects internally, but got converted to and from apps/v1beta1 and apps/v1beta2 Scale objects via conversion (autoConvert_extensions_Scale_To_v1beta1_Scale and autoConvert_extensions_Scale_To_v1beta2_Scale, etc), and those go types actually do exist (the creation of those types was part of the problem, but c'est la vie)

Ah, whoops, forgot that those slipped in and must've misread discovery :-/

DirectXMan12 committed:
[client-go] Add apps.Scale support to Scale client

apps/v1betaX inadventertently contains its own variant of Scale.  In
order to support scaling Deployments, ReplicaSets, etc, we need to support
these versions of Scale as well.

DirectXMan12 commented:
ok, this ^ version has support in the scale client for apps/v1beta{1,2}.  @kubernetes/sig-apps-bugs *please* make sure we don't repeat these mistakes for `apps/v1`.

spiffxp commented:
Getting there!  This looks like a valid failure
```
I1027 22:14:13.058] FAILED   hack/make-rules/../../hack/verify-golint.sh	59s
```

DirectXMan12 commented:
yep, API code doesn't pass golint.

DirectXMan12 commented:
that... doesn't look like this PR failed anything.  Let's see...

/retest

liggitt commented:
fyi, https://github.com/kubernetes/kubernetes/pull/54683 merged... not sure if this needs rebasing on master

DirectXMan12 commented:
/test gci-gce-serial

DirectXMan12 commented:
or do I need

/test ci-kubernetes-e2e-gci-gce-serial

kow3ns commented:
@DirectXMan12 I think we have decided on autoscaling/v1 as the canonical scale sub-resource for v1.

DirectXMan12 commented:
> I think we have decided on autoscaling/v1 as the canonical scale sub-resource for v1.

Good.  That'll be the scale subresource for things going forward.  For everything else, we need to keep compat.

spiffxp commented:
/approve
under the pretense this will fix broken tests

@kubernetes/sig-autoscaling-pr-reviews @kubernetes/api-reviewers 
I don't really feel qualified to /lgtm this, can someone please review?

liggitt commented on test/e2e/common/autoscaling_utils.go:
```diff
@@ -401,28 +400,14 @@ func (rc *ResourceConsumer) CleanUp() {
-	kind, err := kindOf(rc.kind)
-	framework.ExpectNoError(err)
+	kind := rc.kind.GroupKind()
 	framework.ExpectNoError(framework.DeleteResourceAndPods(rc.clientSet, rc.internalClientset, kind, rc.nsName, rc.name))
```
`framework.DeleteResourceAndPods` calls `getRuntimeObjectForKind`, which only recognizes extensions.ReplicaSet and extensions.Deployment. Not sure whether you should add `apps` support to `getRuntimeObjectForKind` (would probably require other changes) or switch your types here to `extensions`

liggitt replied:
Probably need to exercise these types in both the extensions and apps groups, so maybe both :-/

liggitt commented on staging/src/k8s.io/client-go/scale/client_test.go:
```diff
@@ -75,7 +75,7 @@ func fakeScaleClient(t *testing.T) (ScalesGetter, []schema.GroupResource) {
 			GroupVersion: appsv1beta2.SchemeGroupVersion.String(),
```
add an apps/v1beta1 test?

liggitt commented:
@DirectXMan12 not seeing serial tests run against this. based on https://github.com/kubernetes/kubernetes/pull/54586#discussion_r148043176 I wouldn't expect this to work without teaching the framework about apps workload types

DirectXMan12 commented:
>  not seeing serial tests run against this

hmm... yeah.  @kubernetes/test-infra-maintainers am I doing something wrong?  According to the docs on the bot commands, I should be able to use `/test` to trigger any test, right?

krzyzacy commented:
@DirectXMan12 nope we don't allow trigger random ci jobs, or it'll be too easy to DoS us :-) you can only `/test` presubmit runs.

DirectXMan12 commented:
@krzyzacy that makes sense.  Any way we can trigger the serial tests here eventually?  We should have fully confirmation that this actually fixes the original issue.

DirectXMan12 commented:
I don't think I have those permissions.  Let me fix up what @liggitt pointed out, then we'll go from there.

krzyzacy commented:
yeah currently you don't, we'll probably make some auth process so that any k8s member can request rerun of a job

DirectXMan12 commented:
@krzyzacy in the mean time, is there any way you can trigger a serial run to make sure this covers the test cases that aren't normally run as part of the PR jobs (we're mainly looking for the HPA tests).

krzyzacy commented:
@DirectXMan12 CI does not consume things has not been merged yet, I would just run an e2e locally and make sure the affected tests looks green.

DirectXMan12 commented:
@krzyzacy ack.  That works in this case, although I imagine there other are cases where that would be difficult, especially if one does not have easy access to GCE (not all community members work at companies with GCE accounts ;-) ).  We should have the ability to trigger specific sets of tests not normally run on pulls, even if it's just for accountability on PRs like this one.  I'll file a bug.

krzyzacy commented:
@DirectXMan12 yeah, @Q-Lee is working on running e2e against a local cluster, which will make life easier :-)

DirectXMan12 commented:
also, @kubernetes/sig-autoscaling-feature-requests since we now have control over the upscale/downscale intervals, we might be able to move some of these tests off of "slow" by setting those

DirectXMan12 commented:
ok, ran the whole set of HPA tests (after a bit of resource tweaking to get them to run on a one-node-in-a-local-VM cluster).  All the affected parts run fine (full disclosure, 1 one the tests failed on the stability portion, but otherwise look fine).

DirectXMan12 commented:
@liggitt  this look good to go?

liggitt commented:
/lgtm

DirectXMan12 commented:
ping @fabianofranz

fabianofranz commented:
/approve
